// 6pp/src/lib/validators.ts
const isIncludeNumber = (password: string) => {
  const regex = /\d/; // This regular expression matches any digit (0-9)
  return regex.test(password); // This will return true if the password includes a number, and false otherwise
};

const isIncludeSpecialChar = (password: string) => {
  const specialCharRegex = /[ `!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~]/; // This regular expression matches any special character
  return specialCharRegex.test(password); // This will return true if the password includes a special character, and false otherwise
};

const isIncludeCapitalLetter = (password: string) => {
  const capitalLetterRegex = /[A-Z]/; // This regular expression matches any capital letter (A-Z)
  return capitalLetterRegex.test(password); // This will return true if the password includes a capital letter, and false otherwise
};

const isIncludeLowercaseLetter = (password: string) => {
  const lowercaseLetterRegex = /[a-z]/; // This regular expression matches any lowercase letter (a-z)
  return lowercaseLetterRegex.test(password); // This will return true if the password includes a lowercase letter, and false otherwise
};

const isValidEmail = (email: string) => {
  const emailRegex = /\S+@\S+\.\S+/;
  return emailRegex.test(email);
};

const isValidUsername = (username: string) => {
  const usernameRegex = /^[a-zA-Z0-9]+$/;
  return usernameRegex.test(username);
};

const isValidUrl = (url: string) => {
  try {
    new URL(url);
    return true;
  } catch (_) {
    return false;
  }
};

// 10 digit phone number
const isValidPhoneNumber = (phoneNumber: string) => {
  const regex = /^\d{10}$/;
  return regex.test(phoneNumber);
};

// YYYY-MM-DD Format
const isValidDate = (date: string) => {
  const regex = /^\d{4}-\d{2}-\d{2}$/;
  return regex.test(date);
};

// HH:MM Format
const isValidTime = (time: string) => {
  const regex = /^\d{2}:\d{2}$/;
  return regex.test(time);
};

// YYYY-MM-DDTHH:MM Format
const isValidDateTime = (dateTime: string) => {
  const regex = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}$/;
  return regex.test(dateTime);
};

const isDateBetween = (date: string, startDate: string, endDate: string) => {
  const targetDate = new Date(date);
  const start = new Date(startDate);
  const end = new Date(endDate);
  return start <= targetDate && targetDate <= end;
};

export {
  isIncludeNumber,
  isIncludeSpecialChar,
  isIncludeCapitalLetter,
  isIncludeLowercaseLetter,
  isValidEmail,
  isValidUsername,
  isValidUrl,
  isValidPhoneNumber,
  isValidDate,
  isValidTime,
  isValidDateTime,
  isDateBetween,
};
