// 6pp/src/hooks/useRating.tsx
import { useEffect, useState } from "react";
import { UseRatingPropsType } from "../types/hooks.types";
import * as React from "react";

export const useRating = ({
  IconFilled,
  IconOutline,
  maxRating = 5,
  value = 0,
  selectable = false,
  styles = {},
}: UseRatingPropsType) => {
  if (maxRating < 1) throw new Error("maxRating should be greater than 0");

  if (value > maxRating)
    throw new Error("value should be less than Max Rating");

  const [rating, setRating] = useState<number>(value);

  const handleClick = (newRating: number) => {
    if (!selectable) return;
    setRating(newRating);
  };

  useEffect(() => {
    setRating(value);
  }, [value]);

  const Ratings = () => (
    <div
      style={{
        display: "flex",
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "center",
        ...styles,
      }}
    >
      {Array.from({ length: maxRating }).map((_, i) => (
        <span
          style={{
            cursor: selectable ? "pointer" : "unset",
          }}
          key={i}
          onClick={() => handleClick(i + 1)}
        >
          {/* Render your star icon here */}
          {i < rating && <span>{IconFilled}</span>}
          {i >= rating && <span>{IconOutline}</span>}
        </span>
      ))}
    </div>
  );

  return { Ratings, rating, setRating };
};
